sbt client-server port discovery

eed3si9n opened the pull request:
This implements JSON-based port file. Throughout the lifetime of the sbt server, there will be `cwd / "project" / "target" / "active.json"`, which contains `uri` field.

```json
{"uri":"tcp://127.0.0.1:5010"}
```

Using this a potential client, such as IDEs, can find out which port number to hit.

Ref #3508

eed3si9n committed:
implement server handshake test

eed3si9n committed:
JSON port file

This implements JSON-based port file. Thoughout the lifetime of the sbt server there will be `cwd / "project" / "target" / "active.json"`, which contains `url` field.

Using this `url` the potential client, such as IDEs can find out which port number to hit.

Ref #3508

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3556/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

eed3si9n committed:
Fixes test

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3556/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

eed3si9n committed:
Use uri instead of url

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3556/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

eed3si9n committed:
implement tokenfile authentication

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3556/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

eed3si9n committed:
expand the token out to 128-bits

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3556/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/191/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/3556/head-1.x-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/191/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3556">pull/3556/head      </A> | sbt/sbt@252e803de8c89a437f67b7e1120f5f24c511ee61
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | sbt/zinc@32e106d62abd8501863dbaa0d08a53cab4cf64e1
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@62004b28dfcdc33273ebb07a563e78178f6eb21d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@293666f653d2686e3049dce37fb80ccb58845b99
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@cba6f83447b84cc1e967f25a930d4f411f6f0664
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/3556/head&zinc=1.x&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eed3si9n committed:
Reference token file using URI and full file path

Node didn't seem to like read URI out of the box, and I am not sure if File -> URI -> File conversion is universally accepted.

Ref sbt/sbt#3088

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3556/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>
